src/checkpointd/runtime.rs

```diff
@@ -0,0 +1,336 @@
+use crate::farcaster_core::consensus::Encodable;
+use farcaster_core::swap::SwapId;
+use lmdb::{Cursor, Transaction as LMDBTransaction};
+use std::path::PathBuf;
+use std::{
+    any::Any,
+    collections::{HashMap, HashSet},
+    convert::{TryFrom, TryInto},
+    io::{self, Write},
+    ptr::swap_nonoverlapping,
+    str::FromStr,
+};
+use strict_encoding::StrictDecode;
+use strict_encoding::StrictEncode;
+
+use crate::swapd::get_swap_id;
+use crate::Endpoints;
+use crate::LogStyle;
+use bitcoin::hashes::{ripemd160, Hash};
+
+use crate::{
+    rpc::{
+        request::{
+            self, BitcoinAddress, Commit, Keys, MoneroAddress, Msg, Params, Reveal, Token, Tx,
+        },
+        Request, ServiceBus,
+    },
+    syncerd::SweepXmrAddress,
+};
+use crate::{CtlServer, Error, Service, ServiceConfig, ServiceId};
+use colored::Colorize;
+use internet2::TypedEnum;
+use microservices::esb::{self, Handler};
+use request::{LaunchSwap, NodeId};
+
+pub fn run(config: ServiceConfig, data_dir: PathBuf) -> Result<(), Error> {
+    let runtime = Runtime {
+        identity: ServiceId::Checkpoint,
+        checkpoints: CheckpointGetSet::new(data_dir).unwrap(),
+        pending_checkpoint_chunks: map![],
+    };
+
+    Service::run(config, runtime, false)
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+struct CheckpointChunk {
+    pub msg_index: usize,
+    pub serialized_state_chunk: Vec<u8>,
```

Comment:
Using a bit to signal more messages arriving works if the messages are guaranteed to arrive in order. While this is a guarantee that holds here even with our complex architecture in practice, I don't wholly trust it, since some things could go wrong during routing, and think it is more defensive to give each of the chunks an index and an expected total end size.